### PR TITLE
[SPARK-37360][SQL] Support TimestampNTZ in JSON data source

### PR DESCRIPTION
**Pull request by sadikovi**
### What changes were proposed in this pull request?

This PR adds support for TimestampNTZ type in the JSON data source. 

Most of the functionality has already been added, this patch verifies that writes + reads work for TimestampNTZ type and adds schema inference depending on the timestamp value format written. The following applies:
- If there is a mixture of `TIMESTAMP_NTZ` and `TIMESTAMP_LTZ` values, use `TIMESTAMP_LTZ`.
- If there are only `TIMESTAMP_NTZ` values, resolve using the the default timestamp type configured with `spark.sql.timestampType`.

In addition, I introduced a new JSON option `timestampNTZFormat` which is similar to `timestampFormat` but it allows to configure read/write pattern for `TIMESTAMP_NTZ` types. It is basically a copy of timestamp pattern but without timezone.

### Why are the changes needed?

The PR fixes issues when writing and reading TimestampNTZ to and from JSON.

### Does this PR introduce _any_ user-facing change?

Previously, JSON data source would infer timestamp values as `TimestampType` when reading a JSON file. Now, the data source would infer the timestamp value type based on the format (with or without timezone) and default timestamp type based on `spark.sql.timestampType`.

A new JSON option `timestampNTZFormat` is added to control the way values are formatted during writes or parsed during reads.

### How was this patch tested?

I extended `JsonSuite` with a few unit tests to verify that write-read roundtrip works for `TIMESTAMP_NTZ` and `TIMESTAMP_LTZ` values.

### CODE REVIEW

**Comment by gengliangwang** [comment]
I left some comments in the PR for CSV: https://github.com/apache/spark/pull/34596
Let's revisit this one after the CSV one is merged.

**Comment by sadikovi** [comment]
@gengliangwang @MaxGekk could you review the PR? Thank you.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:153]: Could you adjust the comment for inferTimestamp and mention the ntz type:
https://github.com/apache/spark/blob/1235bd29f087db834777cb9fc4c53a12e7f7d6b3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala#L141

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:2751]: It is a dataset/dataframe not an expression, just in case.

Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:2751]: It means "expected", not "expression".

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala
Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:153]: Yes, sure. I will update, thanks for pointing it out!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:2770]: Could you append a fraction part like `timestamp_ltz'2020-12-12 12:12:12.123456'`. Though I would guess JSON ds should write `.000000`, and if something went wrong, it won't be able to read even such zeros. But just in case.

**Review by MaxGekk — APPROVED** [review]
LGTM except of a minor comment.

**Review by gengliangwang — APPROVED** [review]
LGTM if Max's comments are addressed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala
Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:153]: Updated! Could you check the latest PR version? Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: This is a good test case! Updated, thank you.

**Comment by sadikovi** [comment]
JSONOptions have the following comment for `inferTimestamp`: 
> Enables inferring of TimestampType and TimestampNTZType from strings matched to the corresponding timestamp pattern defined by the timestampFormat and timestampNTZFormat options respectively.

Let me know if you would like any modifications to the javadoc. 
I also updated the tests, would appreciate another round of reviews. Thanks!

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @sadikovi and @gengliangwang for review.

**Comment by sadikovi** [comment]
Thank you!

**Pull request closed without merging**
